### ISSUE TITLE
kvserver: support removal of a store from a node

### ISSUE THREAD

**Issue by mzimmerman** [cockroachdb/cockroach#127620]
**Describe the problem**
We have a server that has three disks mounted with --store and the server notated an issue in the logs.
``` Error: pebble/table: invalid table 523484 (checksum mismatch at 57006851/15086)```
I assumed (maybe incorrectly) that the disk was starting to fail but hadn't failed yet, or there was a bug in the underlying write, filesystem, driver, or OS that was triggered.

**To Reproduce**
This disk did not seem to have any hardware level failures, however since I didn't know how to resolve the issue properly (how do you resolve that issue properly?) I removed all the data on that --store and restarted the node including that (now empty) store also.

**Expected behavior**
I expected that cockroach would recognize that a disk had failed and be able to pull replicated copies from the other nodes for the data it was missing.
Since there's a list of multiple stores allowed, certainly this is so that you can lose one store and still continue operating with the stores that exist?

**Additional data / screenshots**
When I started the node back up, there were lots of log messages like:
```W240722 13:39:06.923417 779622 kv/kvserver/raft_transport.go:454 ⋮ [T1,Vsystem,n40] 98191  unable to accept Raft message from (n31,s36):?: no handler registered for (n40,s59):?```
I am guessing that the cluster thought this node had that data, but the node didn't have it and so the messages were getting rejected.  I let this continue for ~10 minutes thinking the cluster would figure out what happened and correct, but the messages continued.

Reading https://www.cockroachlabs.com/docs/stable/recommended-production-settings did not provide clarity.  It states "If you start a node with multiple --store flags, it is not possible to scale back down to only using a single store on the node. Instead, you must decommission the node and start a new node with the updated --store."

- Since it only says you can't scale down to 1 store, can you scale down from N to N-1 stores -- or should it say that you can't scale down the number of stores at all?  Maybe it should more explicitly say that losing one --store on the node loses the whole node?

**Environment:**
```cockroach version details:
Build Tag:        v24.1.1
Build Time:       2024/06/14 07:28:29
Distribution:     CCL
Platform:         linux amd64 (x86_64-pc-linux-gnu)
Go Version:       go1.22.3 X:nocoverageredesign
C Compiler:       gcc 6.5.0
Build Commit ID:  484f8e2f4d3cb23582416944d79eab72ff42a6a1
Build Type:       release
Enabled Assertions: false
```
```Linux OS -- 5.15.0-113-generic #123-Ubuntu SMP Mon Jun 10 08:16:17 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux```

**Additional context**
- Is there any benefit to having multiple --store specified or is it better to just RAID them all together?  I've always thought if you have cluster redundancy, then having redundancy at the disk level is just extra $$ you're wasting.  Why store data redundantly at the disk level if you store it redundantly at the cluster level?
- Ideally, I think replication should only happen at the cluster level, but losing one disk should only lose the data on *that* disk.  If we RAID0 them together, we lose that capability.  If we RAID(1,5,6,10,etc) them together, we lose disk space to "double" replication.

Jira issue: CRDB-40516

Epic CRDB-41111

**Comment by mzimmerman** [comment]
@jbowens I assume from the changing of the title and labels that this is not supported (that's what I thought) so the filing of the report is not necessarily for the enhancement (though I think it'd be awesome) but is instead for fixing the documentation that says:
"If you start a node with multiple --store flags, it is not possible to scale back down to only using a single store on the node. Instead, you must decommission the node and start a new node with the updated --store"

That should be rephrased as it's not true at the moment.  A suggested edit would be:

"If you start a node with multiple --store flags, losing any one of those stores will require the decommissioning of the entire node."

**Comment by jbowens** [comment]
@mzimmerman — I might be misunderstanding, but I think it is true? It says "you must **decommission the node** and start a new node with the updated --store." Decommissioning removes the node and all its stores from the cluster. I'll create a docs issue to make it clearer that decommissioning the node necessarily requires loss of all its stores.

**Comment by mzimmerman** [comment]
Yah, I'm not focusing on the decommissioning part.  It's the "scale back down to only using a single store" part that's the problem.  If it's only a single store that's the issue, if I have 3 stores, I could scale down to 2, but I don't think that's possible either as I wasn't able to when I lost a disk.  So it needs to say you can't scale down at all.  E.g., if you lose one disk (or reconfigure with less disks than prior), you have to decommission the whole node.

**Comment by jbowens** [comment]
Yeah, I think the doc _is_ saying that scaling back is _not_ possible:

"it is **not possible** to scale back down to only using a single store on the node."

**Comment by mzimmerman** [comment]
I realize I can't scale back down to a single node; it is clear on that.  I had thought I'd be able to scale down to N-1 nodes (as long as N didn't equal 2)  So if I configure with 5 disks, losing a disk I should be able to scale down to 4.  The statement does *not* prohibit that which is why I think it's unclear.  If losing *any* disk is not possible to keep the node, I'm not sure why any statement about "single" is important.
